### PR TITLE
[fp8]  support fp8 communication and fp8 training for Colossalai

### PR DESCRIPTION
**Pull request by wangbluo**
## 🚨 Issue number

#5873 #5916 #5926 #5964 #5886 #5960 #5943 
#5927 #5999 #5972 #5974 #5915 #5996 #6004 
#6005 #5955 #5971 

## 📝 What does this PR do?

Supported FP8 communication and training for ColossalAI, including MoeHybridPlugin, HybridPlugin, Zero, and Gemini plugins. The model's loss convergence and accuracy have already tested.

@BurkeHulk @GuangyaoZhang @botbw @ver217 @flybird11111  Please have a look, thank you! :）

### CODE REVIEW

File: colossalai/shardformer/layer/attn.py [hunk 1]
```diff
@@ -165,6 +165,10 @@ def prepare_attn_kwargs(
             if s_q != 1:
                 attention_mask = attention_mask.tril(diagonal=0)
             attention_mask = attention_mask.expand(b, s_q, s_kv)
+            attention_mask = torch.ones(s_q, s_kv, dtype=dtype, device=device)
+            if s_q != 1:
+                attention_mask = attention_mask.tril(diagonal=0)
+            attention_mask = attention_mask.expand(b, s_q, s_kv)
```
Comment (ver217) [inline colossalai/shardformer/layer/attn.py:171]: duplicated?

Comment (wangbluo) [reply colossalai/shardformer/layer/attn.py:171]: https://github.com/hpcaitech/ColossalAI/pull/6033

**Review by ver217 — APPROVED** [review]
